File: flink-tests/src/test/scala/org/apache/flink/api/scala/operators/SumMinMaxITCase.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.api.scala.operators

import org.apache.flink.api.scala.ExecutionEnvironment
import org.apache.flink.api.scala.util.CollectionDataSets
import org.apache.flink.test.util.MultipleProgramsTestBase.TestExecutionMode
import org.apache.flink.test.util.MultipleProgramsTestBase

import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.Parameterized
import org.junit.Assert._

import org.apache.flink.api.scala._

@RunWith(classOf[Parameterized])
class SumMinMaxITCase(mode: TestExecutionMode) extends MultipleProgramsTestBase(mode) {

  @Test
  def testFullAggregate(): Unit = {
    // Full aggregate
    val env = ExecutionEnvironment.getExecutionEnvironment
    val ds = CollectionDataSets.get3TupleDataSet(env)

    val aggregateDs : DataSet[(Int, Long)] = ds
      .sum(0)
      .andMax(1)
      // Ensure aggregate operator correctly copies other fields
      .filter(_._3 != null)
      .map{ t => (t._1, t._2) }


    val result: Seq[(Int, Long)] = aggregateDs.collect()

    assertEquals(1, result.size)
    assertEquals((231, 6L), result.head)
  }

  @Test
  def testGroupedAggregate(): Unit = {
    // Grouped aggregate
    val env = ExecutionEnvironment.getExecutionEnvironment
    val ds = CollectionDataSets.get3TupleDataSet(env)

    val aggregateDs = ds
      .groupBy(1)
      .sum(0)
      // Ensure aggregate operator correctly copies other fields
      .filter(_._3 != null)
      .map { t => (t._2, t._1) }

    val result : Seq[(Long, Int)] = aggregateDs.collect().sortWith((a, b) => a._1 < b._1)

    val expected : Seq[(Long, Int)] = Seq((1L, 1), (2L, 5), (3L, 15), (4L, 34), (5L, 65), (6L, 111))
    assertEquals(expected, result)
  }

  @Test
  def testNestedAggregate(): Unit = {
    // Nested aggregate
    val env = ExecutionEnvironment.getExecutionEnvironment
    val ds = CollectionDataSets.get3TupleDataSet(env)

    val aggregateDs: DataSet[Int] = ds
      .groupBy(1)
      .min(0)
      .min(0)
      // Ensure aggregate operator correctly copies other fields
      .filter(_._3 != null)
      .map { t => t._1 }

    val result: Seq[Int] = aggregateDs.collect()

    assertEquals(1, result.size)
    assertEquals(Seq(1), result)
  }
}

File: flink-tests/src/test/java/org/apache/flink/test/operators/SumMinMaxITCase.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.test.operators;

import org.apache.flink.api.java.DataSet;
import org.apache.flink.api.java.ExecutionEnvironment;
import org.apache.flink.api.java.tuple.Tuple1;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.api.java.tuple.Tuple3;
import org.apache.flink.test.operators.util.CollectionDataSets;
import org.apache.flink.test.util.MultipleProgramsTestBase;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;

import java.util.List;

/**
 * Integration tests for {@link org.apache.flink.api.scala.GroupedDataSet#min} and {@link
 * org.apache.flink.api.scala.GroupedDataSet#max}.
 */
@RunWith(Parameterized.class)
public class SumMinMaxITCase extends MultipleProgramsTestBase {

    public SumMinMaxITCase(TestExecutionMode mode) {
        super(mode);
    }

    @Test
    public void testSumMaxAndProject() throws Exception {
        final ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();

        DataSet<Tuple3<Integer, Long, String>> ds = CollectionDataSets.get3TupleDataSet(env);
        DataSet<Tuple2<Integer, Long>> sumDs = ds.sum(0).andMax(1).project(0, 1);

        List<Tuple2<Integer, Long>> result = sumDs.collect();

        String expected = "231,6\n";

        compareResultAsTuples(result, expected);
    }

    @Test
    public void testGroupedAggregate() throws Exception {
        /*
         * Grouped Aggregate
         */

        final ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();

        DataSet<Tuple3<Integer, Long, String>> ds = CollectionDataSets.get3TupleDataSet(env);
        DataSet<Tuple2<Long, Integer>> aggregateDs = ds.groupBy(1).sum(0).project(1, 0);

        List<Tuple2<Long, Integer>> result = aggregateDs.collect();

        String expected = "1,1\n" + "2,5\n" + "3,15\n" + "4,34\n" + "5,65\n" + "6,111\n";

        compareResultAsTuples(result, expected);
    }

    @Test
    public void testNestedAggregate() throws Exception {
        /*
         * Nested Aggregate
         */

        final ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();

        DataSet<Tuple3<Integer, Long, String>> ds = CollectionDataSets.get3TupleDataSet(env);
        DataSet<Tuple1<Integer>> aggregateDs = ds.groupBy(1).min(0).min(0).project(0);

        List<Tuple1<Integer>> result = aggregateDs.collect();

        String expected = "1\n";

        compareResultAsTuples(result, expected);
    }
}
